bakery/src/tasks/upload-book.js

```diff
@@ -38,6 +40,7 @@ const task = ({ awsAccessKeyId, awsSecretAccessKey, bucketName }) => {
           for jsonfile in "$book_dir/"*@*.json; do cp "$jsonfile" "$target_dir/$book_uuid@$book_version:$(basename $jsonfile)"; done;
           for xhtmlfile in "$book_dir/"*@*.xhtml; do cp "$xhtmlfile" "$target_dir/$book_uuid@$book_version:$(basename $xhtmlfile)"; done;
           aws s3 cp --recursive "$target_dir" s3://${bucketName}/contents
+          python /code/scripts/copy-resources-s3.py "$resources_dir" ${bucketName} resources
```

Comment:
Just to confirm my own understanding: The main reason we are wrapping the upload in a custom script using `boto` instead of using the `aws cp / sync` CLI with `--content-type` is because the mime type value is in the JSON metadata file generated and we can't easily just upload all of a single type. Is that right?

Reply:
That is correct. When I think about it now:
- Another approach possible is to use `aws cp` with `--content-type` but let a script generate what file lists with all jpeg, all png etc. for eatch content type.
I wonder if this speeds up the upload process. The downside with the script currently is that it is blocking on each upload which makes the upload process not the fastest. I was thinking of optimizing it with a small number of threads.
- Yet another approach is to upload everything with no content-type with `aws cp` and let the script only correct the content-type on the binary files in s3. I think that would be also faster than the current solution.

I have the feeling this requires some discussion. What do you think is the best in your opinion?

Reply:
Update: The script has also another benefit. It checks (and verifies!) if same content already existing and does not upload it again. AFAIK `aws cp` does not have the same logic but I can be mistaken. I'm leaning of adding threads to the upload script to speed it a little bit up.

Reply:
I don't know if it helps in the decision process, but I think this somewhat depends on how much upload control / customization capabilities we may foresee in the future for the resource files. Per your examples above, right now we get a high degree of flexibility and just need to perhaps improve the performance. On the other hand, if we know things are fairly limited in scope and the CLI saves us from having to maintain some code by using `cp --recursive` or `sync`, your ideas on ways to use that may be worth exploring. 

I'm also not sure how much we want to optimize for the double upload case (e.g. the `cp --recursive` versus `sync` in the CLI scenario) since, ideally, we never run the same distribution version twice.

Reply:
@therealmarv My vote would be to use async io over threads, since this is exactly the type of work it's meant to do best :smile:

Reply:
@therealmarv My vote would be to use async io over threads, since this is exactly the type of work it's meant to do best :smile: